contracts/rollover/CrossCurrencyRollover.sol

```diff
@@ -0,0 +1,415 @@
+     * @param lender                 The address of the new lender.
+     * @param swapParams             The parameters for the currency swap.
+     * @param newLoanTerms           The terms of the new loan.
+     * @param oldLoanData            The loan data of the original loan.
+     * @param oldLoanId              The ID of the original loan to be rolled over.
+     *
+     * @return repayAmount           The amount to be repaid to the old loan.
+     */
+    function _processSettlement(
+        address borrower,
+        address lender,
+        OriginationLibrary.SwapParameters memory swapParams,
+        LoanLibrary.LoanTerms memory newLoanTerms,
+        LoanLibrary.LoanData memory oldLoanData,
+        uint256 oldLoanId
+    ) internal returns (uint256 repayAmount){
+        // pull funds from the new lender
+        IERC20(newLoanTerms.payableCurrency).safeTransferFrom(
+            lender,
+            address(this),
+            newLoanTerms.principal
+        );
+
+        // swap the new lender payment from new currency to the original currency
+        // the swap will revert if swappedAmount < minAmountOut
+        uint256 swappedAmount = _swapExactInputSingle(
+            newLoanTerms.payableCurrency,
+            oldLoanData.terms.payableCurrency,
+            newLoanTerms.principal,
+            swapParams.minAmountOut,
+            swapParams.poolFeeTier,
+            address(this)
+        );
+
+        address oldLender = ILoanCore(loanCore).lenderNote().ownerOf(oldLoanId);
+
+        // calculate settle amounts
+        // get interest amount due
+        uint256 interestAmount = InterestCalculator.getProratedInterestAmount(
+            oldLoanData.balance,
+            oldLoanData.terms.interestRate,
+            oldLoanData.terms.durationSecs,
+            oldLoanData.startDate,
+            oldLoanData.lastAccrualTimestamp,
+            block.timestamp
+        );
+
+        // calculate the repay amount to settle the original loan
+        repayAmount = oldLoanData.terms.principal + interestAmount;
+
+        OriginationLibrary.RolloverAmounts memory amounts = rolloverAmounts(
+            oldLoanData.terms.principal,
+            interestAmount,
+            swappedAmount,
+            lender,
+            oldLender,
+            0,
+            0
+        );
+
+        // borrower owes
+        if (amounts.needFromBorrower > 0) {
+            IERC20(oldLoanData.terms.payableCurrency).safeTransferFrom(borrower, address(this), amounts.needFromBorrower);
+        }
+
+        // amount = original loan repayment amount - leftover principal
+        uint256 amount = amounts.amountFromLender - amounts.leftoverPrincipal;
+
+        if (swappedAmount > amount) {
+            uint256 remainingFunds = swappedAmount - amount;
+
+            // if funds remain, send to the borrower
+            IERC20(oldLoanData.terms.payableCurrency).safeTransfer(borrower, remainingFunds);
+        }
```

Comment:
I still don't know what these lines are doing - is there a better variable name than `amount`?

The comment also confuses things: why does `original loan repayment amount` describe `amounts.amountFromLender`? `amounts.amountFromLender` is the principal for the new loan, not the loan repayment amount.

I get what you want to do here, you want to calculate if funds are leftover, and if so, send them to the borrower. But the name of `amounts.amountToBorrower` suggests exactly that concept, no?  We designed `rolloverAmounts` to calculate exactly things like this (what funds need to go where for settlement), but now we are not using the outputs of that function, instead we're using `swappedAmount`, but that was already an input for `rolloverAmounts` calculation? Since the swap is already done, `rolloverAmounts` should have all the information it needs. 

So, the variable naming, usage of variables, and comments still don't quite clearly describe what is going on here. And as regards the accounting itself, if we break down the math in `OriginationCalculator`, we can see that:

- `amounts.amountFromLender == newPrincipalAmount`
- For the case where the borrower has funds leftover, so `repayAmount < newPrincipalAmount`, then `amounts.leftoverPrincipal == newPrincipalAmount + totalFees - repayAmount`

So, `amounts.amountFromLender - amounts.leftoverPrincipal` is equivalent to `newPrincipalAmount - newPrincipalAmount + totalFees - repayAmount`, which is equivalent to `totalFees - repayAmount`. Is this actually what we want to send to the borrower?

I recommend writing a test script where fees are turned onto some reasonable value (e.g. a 10% interest fee). I think the accounting might become more clear if we had those turned on. And my intuition is that, if we need to figure out what to send to the borrower, we should use `amounts.amountToBorrower`. It's what it's named for, after all.